File: app/indicators.py
def ema(values: list[float], period: int) -> list[float]:
    if period <= 0:
        raise ValueError("period must be positive")
    if len(values) < period:
        return []

    multiplier = 2 / (period + 1)
    seed = sum(values[:period]) / period
    results = [seed]

    for price in values[period:]:
        next_value = ((price - results[-1]) * multiplier) + results[-1]
        results.append(next_value)

    return results


def rsi(values: list[float], period: int = 14) -> float:
    if len(values) <= period:
        return 50.0

    gains: list[float] = []
    losses: list[float] = []

    for i in range(1, len(values)):
        delta = values[i] - values[i - 1]
        if delta >= 0:
            gains.append(delta)
            losses.append(0.0)
        else:
            gains.append(0.0)
            losses.append(abs(delta))

    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period

    for i in range(period, len(gains)):
        avg_gain = ((avg_gain * (period - 1)) + gains[i]) / period
        avg_loss = ((avg_loss * (period - 1)) + losses[i]) / period

    if avg_loss == 0:
        return 100.0

    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def macd(values: list[float], fast: int = 12, slow: int = 26, signal: int = 9) -> tuple[float, float]:
    if len(values) < slow + signal:
        return 0.0, 0.0

    fast_ema = ema(values, fast)
    slow_ema = ema(values, slow)

    if not fast_ema or not slow_ema:
        return 0.0, 0.0

    offset = len(fast_ema) - len(slow_ema)
    aligned_fast = fast_ema[offset:]
    macd_line = [f - s for f, s in zip(aligned_fast, slow_ema)]

    signal_line_values = ema(macd_line, signal)
    if not signal_line_values:
        return macd_line[-1], 0.0

    return macd_line[-1], signal_line_values[-1]

File: lib/indicators.ts
export function ema(values: number[], period: number): number[] {
  if (period <= 0 || values.length < period) return [];

  const multiplier = 2 / (period + 1);
  const seed = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  const out = [seed];

  for (const price of values.slice(period)) {
    out.push((price - out[out.length - 1]) * multiplier + out[out.length - 1]);
  }

  return out;
}

export function rsi(values: number[], period = 14): number {
  if (values.length <= period) return 50;

  const gains: number[] = [];
  const losses: number[] = [];

  for (let i = 1; i < values.length; i += 1) {
    const delta = values[i] - values[i - 1];
    gains.push(delta > 0 ? delta : 0);
    losses.push(delta < 0 ? Math.abs(delta) : 0);
  }

  let avgGain = gains.slice(0, period).reduce((a, b) => a + b, 0) / period;
  let avgLoss = losses.slice(0, period).reduce((a, b) => a + b, 0) / period;

  for (let i = period; i < gains.length; i += 1) {
    avgGain = (avgGain * (period - 1) + gains[i]) / period;
    avgLoss = (avgLoss * (period - 1) + losses[i]) / period;
  }

  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function macd(values: number[], fast = 12, slow = 26, signalPeriod = 9): { macd: number; signal: number } {
  if (values.length < slow + signalPeriod) return { macd: 0, signal: 0 };

  const fastEma = ema(values, fast);
  const slowEma = ema(values, slow);
  if (!fastEma.length || !slowEma.length) return { macd: 0, signal: 0 };

  const offset = fastEma.length - slowEma.length;
  const alignedFast = fastEma.slice(offset);
  const macdLine = alignedFast.map((v, i) => v - slowEma[i]);
  const signalLine = ema(macdLine, signalPeriod);

  return {
    macd: macdLine[macdLine.length - 1] ?? 0,
    signal: signalLine[signalLine.length - 1] ?? 0,
  };
}
